File: src/cost/tracker.ts
import type { Usage } from "../types.ts";

export interface TurnUsage {
	promptTokens: number;
	completionTokens: number;
	totalTokens: number;
	cost: number | null;
	cachedTokens?: number;
	reasoningTokens?: number;
	timestamp: string;
}

export class CostTracker {
	private turns: TurnUsage[] = [];

	addUsage(usage: Usage): void {
		const turn: TurnUsage = {
			promptTokens: usage.prompt_tokens,
			completionTokens: usage.completion_tokens,
			totalTokens: usage.total_tokens,
			cost: usage.cost ?? null,
			cachedTokens: usage.prompt_tokens_details?.cached_tokens,
			reasoningTokens: usage.completion_tokens_details?.reasoning_tokens,
			timestamp: new Date().toISOString(),
		};
		this.turns.push(turn);
	}

	get totalInputTokens(): number {
		return this.turns.reduce((sum, t) => sum + t.promptTokens, 0);
	}

	get totalOutputTokens(): number {
		return this.turns.reduce((sum, t) => sum + t.completionTokens, 0);
	}

	get totalCost(): number | null {
		const withCost = this.turns.filter((t) => t.cost !== null);
		if (withCost.length === 0) return null;
		return withCost.reduce((sum, t) => sum + (t.cost as number), 0);
	}

	get breakdown(): readonly TurnUsage[] {
		return this.turns;
	}

	isOverBudget(limit: number): boolean {
		const cost = this.totalCost;
		return cost !== null && cost > limit;
	}

	reset(): void {
		this.turns = [];
	}
}

File: src/cost/tracker.rs
//! Accumulates per-turn token usage and computes totals.

use chrono::Utc;

use crate::types::Usage;

#[derive(Debug, Clone)]
pub struct TurnUsage {
    pub prompt_tokens: u64,
    pub completion_tokens: u64,
    pub total_tokens: u64,
    pub cost: Option<f64>,
    pub cached_tokens: Option<u64>,
    pub reasoning_tokens: Option<u64>,
    pub timestamp: String,
}

#[derive(Debug, Default, Clone)]
pub struct CostTracker {
    turns: Vec<TurnUsage>,
}

impl CostTracker {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_usage(&mut self, usage: &Usage) {
        self.turns.push(TurnUsage {
            prompt_tokens: usage.prompt_tokens,
            completion_tokens: usage.completion_tokens,
            total_tokens: usage.total_tokens,
            cost: usage.cost,
            cached_tokens: usage
                .prompt_tokens_details
                .as_ref()
                .and_then(|d| d.cached_tokens),
            reasoning_tokens: usage
                .completion_tokens_details
                .as_ref()
                .and_then(|d| d.reasoning_tokens),
            timestamp: Utc::now().to_rfc3339(),
        });
    }

    pub fn total_input_tokens(&self) -> u64 {
        self.turns.iter().map(|t| t.prompt_tokens).sum()
    }

    pub fn total_output_tokens(&self) -> u64 {
        self.turns.iter().map(|t| t.completion_tokens).sum()
    }

    pub fn total_cost(&self) -> Option<f64> {
        let with_cost: Vec<f64> = self.turns.iter().filter_map(|t| t.cost).collect();
        if with_cost.is_empty() {
            None
        } else {
            Some(with_cost.iter().sum())
        }
    }

    pub fn breakdown(&self) -> &[TurnUsage] {
        &self.turns
    }

    pub fn is_over_budget(&self, limit: f64) -> bool {
        self.total_cost().map(|c| c > limit).unwrap_or(false)
    }

    pub fn reset(&mut self) {
        self.turns.clear();
    }
}
